storage/sector_states.go

```diff
@@ -0,0 +1,231 @@
+	}
+
+	fillerSizes, err := fillersFromRem(ubytes - allocated)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(fillerSizes) > 0 {
+		log.Warnf("Creating %d filler pieces for sector %d", len(fillerSizes), sector.SectorID)
+	}
+
+	pieces, err := m.storeGarbage(ctx, sector.SectorID, sector.existingPieces(), fillerSizes...)
+	if err != nil {
+		return nil, xerrors.Errorf("filling up the sector (%v): %w", fillerSizes, err)
+	}
+
+	return func(info *SectorInfo) {
+		info.Pieces = append(info.Pieces, pieces...)
+	}, nil
+}
+
+func (m *Miner) sealPreCommit(ctx context.Context, sector SectorInfo) (func(*SectorInfo), error) {
+	log.Infow("performing sector replication...", "sector", sector.SectorID)
+	ticket, err := m.tktFn(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	rspco, err := m.sb.SealPreCommit(sector.SectorID, *ticket, sector.pieceInfos())
+	if err != nil {
+		return nil, xerrors.Errorf("seal pre commit failed: %w", err)
+	}
+
+	return func(info *SectorInfo) {
+		info.CommC = rspco.CommC[:]
+		info.CommD = rspco.CommD[:]
+		info.CommR = rspco.CommR[:]
+		info.CommRLast = rspco.CommRLast[:]
+		info.Ticket = SealTicket{
+			BlockHeight: ticket.BlockHeight,
+			TicketBytes: ticket.TicketBytes[:],
+		}
+	}, nil
+}
+
+func (m *Miner) preCommit(ctx context.Context, sector SectorInfo) (func(*SectorInfo), error) {
+	params := &actors.SectorPreCommitInfo{
+		SectorNumber: sector.SectorID,
+
+		CommR:     sector.CommR,
+		SealEpoch: sector.Ticket.BlockHeight,
+		DealIDs:   sector.deals(),
+	}
+	enc, aerr := actors.SerializeParams(params)
+	if aerr != nil {
+		return nil, xerrors.Errorf("could not serialize commit sector parameters: %w", aerr)
+	}
+
+	msg := &types.Message{
+		To:       m.maddr,
+		From:     m.worker,
+		Method:   actors.MAMethods.PreCommitSector,
+		Params:   enc,
+		Value:    types.NewInt(0), // TODO: need to ensure sufficient collateral
+		GasLimit: types.NewInt(1000000 /* i dont know help */),
+		GasPrice: types.NewInt(1),
+	}
+
+	log.Info("submitting precommit for sector: ", sector.SectorID)
+	smsg, err := m.api.MpoolPushMessage(ctx, msg)
+	if err != nil {
+		return nil, xerrors.Errorf("pushing message to mpool: %w", err)
+	}
+
+	return func(info *SectorInfo) {
+		mcid := smsg.Cid()
+		info.PreCommitMessage = &mcid
+	}, nil
+}
+
+func (m *Miner) preCommitted(ctx context.Context, sector SectorInfo) (func(*SectorInfo), error) {
+	// would be ideal to just use the events.Called handler, but it wouldnt be able to handle individual message timeouts
+	log.Info("Sector precommitted: ", sector.SectorID)
+	mw, err := m.api.StateWaitMsg(ctx, *sector.PreCommitMessage)
+	if err != nil {
+		return nil, err
+	}
+
+	if mw.Receipt.ExitCode != 0 {
+		log.Error("sector precommit failed: ", mw.Receipt.ExitCode)
+		return nil, err
+	}
+	log.Info("precommit message landed on chain: ", sector.SectorID)
+
+	randHeight := mw.TipSet.Height() + build.InteractivePoRepDelay - 1 // -1 because of how the messages are applied
```

Comment:
the -1 seems like an odd constraint. We should probably figure that out and remove it to keep the math clean

Reply:
The correct thing would be to make the `vmctx.BlockHeight()` be the height at which the message is computed instead of the height at which it it included. That's probably tricky, and a bit out of scope here.